File: #006 Luhn - checksum formula/Luhn.py
import sys

def luhn(input):
    sum = 0
    for i in range(len(input)):
        digit = ord(input[i]) - ord('0')
        if (i % 2) != 0:
            digit = digit * 2
            if digit > 9:
                digit = digit - 9
        sum += digit
    return (sum % 10 == 0)

def main():
    input = list(sys.argv[1])
    valid = luhn(input)
    length = len(input)
    firstTwoDigits = int("".join(input[0:2]))

    if valid and length == 15 and ( firstTwoDigits == 37 or firstTwoDigits == 34):
        print("AMEX\n")
    
    elif valid and length == 16 and (firstTwoDigits >= 50 or firstTwoDigits <= 55):
        print("MASTERCARD\n")

    elif valid and (length == 16 or length == 13) and (int(input[0]) == 4):
        print("VISA\n")

    else:
        print("INVALID\n")

if __name__ == "__main__":
    main()

File: #006 Luhn - checksum formula/Luhn.c
// creditcard validation using Luhns algorithm
#include <stdio.h>
#include <stdlib.h>

int luhnAlgorithm(char creditNumb[]){
    int sum = 0;
    for(int i=0; creditNumb[i]!='\0'; i++){
        int digit = (creditNumb[i] - '0');
        if(i % 2 != 0){
            digit = digit * 2;
            if(digit > 9){
                digit = digit - 9;
            }
        }
        sum+=digit;
    }
    return (sum % 10 == 0);
}

int getSizeOfCharArray(char input[]){
    int length = 0;
    for(int i=0; input[i]!='\0'; i++){
        length++;
    }
    return length;
}

int main(){
    printf("Enter credit card number:\n");
    char input[256];
    int length;
    do{
        scanf("%s", &input);
        length = getSizeOfCharArray(input);
        if(length < 13){
            printf("%s", "length must be at least 13\n");
        }
    }while(length < 13);

    char firstTwoDigit[2];
    firstTwoDigit[0] = input[0];
    firstTwoDigit[1] = input[1];
    char *ptr;
    long res = strtol(firstTwoDigit, &ptr, 10);

    int valid = luhnAlgorithm(input);
    if(valid && length == 15 && (res == 37 || res == 34)){
        printf("AMEX\n");
    }
    else if(valid && length == 16 && (res >= 50 || res <= 55)){
        printf("MASTERCARD\n");
    }
    else if(valid && (length == 16 || length == 13) && (firstTwoDigit[0]-'\0') == 4){
        printf("VISA\n");
    }
    else{
        printf("INVALID\n");
    }

    return 0;
}